### PR TITLE
Dependent assets in asset group

### PR DESCRIPTION
**Pull request by padmaia**
# ↪️ Pull Request
With this PR assets can now define dependent assets in the same asset group by declaring a dependency with a module specifer that matches the dependent asset's unique key. These dependent assets are then connected to the asset's dependency instead of the asset group in the AssetGraph.

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.02s | +67.00ms |
| Cached | 4.32s | -110.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 202.00ms | -26.00ms 🚀 |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 201.00ms | -26.00ms 🚀 |
| dist/legacy/index.js | 520.00b | +0.00b | 560.00ms | +33.00ms ⚠️ |
| dist/modern/index.js | 453.00b | +0.00b | 554.00ms | +33.00ms ⚠️ |
| dist/modern/index.css | 29.00b | +0.00b | 1.37s | +119.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 14.00ms | -315.00ms 🚀 |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 13.00ms | -314.00ms 🚀 |
| dist/legacy/index.js | 520.00b | +0.00b | 23.00ms | -312.00ms 🚀 |
| dist/modern/index.js | 453.00b | +0.00b | 14.00ms | -315.00ms 🚀 |
| dist/legacy/index.css | 29.00b | +0.00b | 12.00ms | -315.00ms 🚀 |
| dist/modern/index.css | 29.00b | +0.00b | 12.00ms | -315.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 12.83s | +663.00ms ⚠️ |
| Cached | 4.46s | -31.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/module/index.js | 34.29kb | +0.00b | 10.00ms | -329.00ms 🚀 |
| dist/main/index.js | 34.27kb | +0.00b | 16.00ms | -329.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/ak-editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.55m | +1.12s |
| Cached | 6.85s | -0.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 2.33mb | +6.00b ⚠️ | 2.03m | -625.00ms |
| dist/pdfRenderer.85e56a51.js | 1.11mb | +0.00b | 42.57s | +5.91s ⚠️ |
| dist/EmojiPickerComponent.7202df5c.js | 105.88kb | +0.00b | 53.20s | +7.24s ⚠️ |
| dist/Toolbar.5c7a6fbb.js | 105.61kb | +0.00b | 1.44m | +5.36s ⚠️ |
| dist/ui.ab1a2648.js | 73.70kb | +0.00b | 1.13m | +6.44s ⚠️ |
| dist/smartMediaEditor.e65bd2ee.js | 67.39kb | +0.00b | 1.13m | +6.44s ⚠️ |
| dist/component.10b390dc.js | 46.17kb | +0.00b | 46.04s | +5.53s ⚠️ |
| dist/DatePicker.0242ec52.js | 27.57kb | +0.00b | 53.52s | +8.81s ⚠️ |
| dist/EmojiPickerComponent.e7c7bb11.js | 26.74kb | +0.00b | 51.61s | +6.90s ⚠️ |
| dist/popup.4290aef9.js | 22.34kb | +0.00b | 44.74s | +5.34s ⚠️ |
| dist/popup.5b4d772e.js | 21.80kb | +0.00b | 44.74s | +5.33s ⚠️ |
| dist/popup.a791635d.js | 20.82kb | +0.00b | 44.74s | +5.34s ⚠️ |
| dist/popup.e0094d89.js | 12.77kb | +0.00b | 44.74s | +5.34s ⚠️ |
| dist/workerHasher.d7e20c98.js | 11.56kb | +0.00b | 54.26s | +7.58s ⚠️ |
| dist/EmojiTypeAheadComponent.1c09c30d.js | 10.44kb | +0.00b | 53.98s | +7.58s ⚠️ |
| dist/card.f99e1597.js | 5.89kb | +0.00b | 16.75s | +1.71s ⚠️ |
| dist/component.a26e593f.js | 5.62kb | +0.00b | 46.34s | +5.53s ⚠️ |
| dist/index.css | 3.38kb | +0.00b | 1.46m | +5.47s ⚠️ |
| dist/component.8fa44754.js | 3.38kb | +0.00b | 46.34s | +5.53s ⚠️ |
| dist/EmojiUploadComponent.0460cc59.js | 2.91kb | +0.00b | 53.99s | +7.59s ⚠️ |
| dist/component.ab65829f.js | 2.29kb | +0.00b | 46.34s | +5.53s ⚠️ |
| dist/media-picker-analytics-error-boundary.0f11d632.js | 1.85kb | +0.00b | 45.38s | +5.44s ⚠️ |
| dist/card.2c902906.js | 1.73kb | +0.00b | 16.75s | +1.71s ⚠️ |
| dist/16.dcacbe5c.js | 1.69kb | +0.00b | 46.85s | +5.55s ⚠️ |
| dist/date.cfe6153f.js | 1.63kb | +0.00b | 47.11s | +5.56s ⚠️ |
| dist/16.de3f3184.js | 1.62kb | +0.00b | 46.52s | +5.55s ⚠️ |
| dist/images.c60a97a9.js | 1.57kb | +0.00b | 47.12s | +5.57s ⚠️ |
| dist/feedback.01d7bcf7.js | 1.53kb | +0.00b | 51.61s | +7.42s ⚠️ |
| dist/workerHasher.854262ce.js | 1.48kb | +0.00b | 53.99s | +7.30s ⚠️ |
| dist/status.e3ed4764.js | 1.35kb | +0.00b | 47.47s | +5.61s ⚠️ |
| dist/list-number.dbe744df.js | 1.35kb | +0.00b | 47.24s | +5.59s ⚠️ |
| dist/16.f54dac28.js | 1.34kb | +0.00b | 46.67s | +5.54s ⚠️ |
| dist/16.8f0e8fe1.js | 1.29kb | +0.00b | 46.52s | +5.55s ⚠️ |
| dist/16.43b62c01.js | 1.29kb | +0.00b | 46.53s | +5.56s ⚠️ |
| dist/code.092e0137.js | 1.28kb | +0.00b | 47.05s | +5.61s ⚠️ |
| dist/16.b17a4d30.js | 1.27kb | +0.00b | 47.01s | +5.57s ⚠️ |
| dist/16.51bbc68e.js | 1.23kb | +0.00b | 46.90s | +5.61s ⚠️ |
| dist/component.8351f228.js | 1.23kb | +0.00b | 45.38s | +5.44s ⚠️ |
| dist/link.6fa27bb2.js | 1.20kb | +0.00b | 47.24s | +5.59s ⚠️ |
| dist/heading6.3996fe48.js | 1.19kb | +0.00b | 47.57s | +5.58s ⚠️ |
| dist/16.385d3ff5.js | 1.19kb | +0.00b | 46.85s | +5.55s ⚠️ |
| dist/heading3.9fea5413.js | 1.18kb | +0.00b | 47.56s | +5.57s ⚠️ |
| dist/16.b3ce4050.js | 1.16kb | +0.00b | 46.52s | +5.55s ⚠️ |
| dist/16.e37bbe6b.js | 1.15kb | +0.00b | 46.85s | +5.55s ⚠️ |
| dist/16.f7fe274a.js | 1.14kb | +0.00b | 47.01s | +5.57s ⚠️ |
| dist/emoji.c4bc5355.js | 1.12kb | +0.00b | 47.11s | +5.56s ⚠️ |
| dist/16.ac2ff83f.js | 1.10kb | +0.00b | 46.72s | +5.58s ⚠️ |
| dist/16.4c652b29.js | 1.10kb | +0.00b | 46.67s | +5.54s ⚠️ |
| dist/16.24165259.js | 1.09kb | +0.00b | 46.52s | +5.55s ⚠️ |
| dist/16.74510c56.js | 1.09kb | +0.00b | 46.67s | +5.54s ⚠️ |
| dist/16.97c5bdf9.js | 1.09kb | +0.00b | 46.67s | +5.54s ⚠️ |
| dist/16.a79db876.js | 1.09kb | +0.00b | 46.39s | +5.56s ⚠️ |
| dist/heading5.2f045997.js | 1.06kb | +0.00b | 47.55s | +5.57s ⚠️ |
| dist/16.6f8ac62a.js | 1.05kb | +0.00b | 46.34s | +5.53s ⚠️ |
| dist/expand.13380081.js | 1.05kb | +0.00b | 51.61s | +7.42s ⚠️ |
| dist/card.aae6ebb9.js | 1.04kb | +0.00b | 16.75s | +1.71s ⚠️ |
| dist/16.3e36da0b.js | 1.01kb | +0.00b | 46.85s | +5.55s ⚠️ |
| dist/heading2.2194210a.js | 1.00kb | +0.00b | 47.55s | +5.57s ⚠️ |
| dist/mention.9fdad24e.js | 981.00b | +0.00b | 47.26s | +5.62s ⚠️ |
| dist/fallback.305c8864.js | 980.00b | +0.00b | 47.47s | +5.61s ⚠️ |
| dist/heading4.63d1ca06.js | 976.00b | +0.00b | 47.55s | +5.57s ⚠️ |
| dist/layout.6f66c2da.js | 962.00b | +0.00b | 47.24s | +5.59s ⚠️ |
| dist/divider.160f3108.js | 940.00b | +0.00b | 47.11s | +5.56s ⚠️ |
| dist/quote.88bfdca4.js | 936.00b | +0.00b | 47.47s | +5.61s ⚠️ |
| dist/action.146450ba.js | 919.00b | +0.00b | 47.01s | +5.57s ⚠️ |
| dist/decision.06254b93.js | 898.00b | +0.00b | 47.11s | +5.56s ⚠️ |
| dist/panel-warning.59c8ef3f.js | 891.00b | +0.00b | 47.36s | +5.61s ⚠️ |
| dist/list.4f623de3.js | 870.00b | +0.00b | 47.24s | +5.59s ⚠️ |
| dist/heading1.65fb8482.js | 864.00b | +0.00b | 47.49s | +5.62s ⚠️ |
| dist/panel-error.c2479698.js | 791.00b | +0.00b | 47.36s | +5.61s ⚠️ |
| dist/panel.23cd7224.js | 788.00b | +0.00b | 47.38s | +5.63s ⚠️ |
| dist/table.4ace5843.js | 780.00b | +0.00b | 47.47s | +5.61s ⚠️ |
| dist/media-viewer-analytics-error-boundary.dd752a9e.js | 777.00b | +0.00b | 16.75s | +1.71s ⚠️ |
| dist/panel-success.6a7b97d7.js | 728.00b | +0.00b | 47.36s | +5.61s ⚠️ |
| dist/panel-note.5ff85bdc.js | 727.00b | +0.00b | 47.36s | +5.61s ⚠️ |
| dist/simpleHasher.884d4eb6.js | 446.00b | +0.00b | 54.29s | +7.58s ⚠️ |
| dist/dropzone.976debcc.js | 177.00b | +1.00b ⚠️ | 2.06m | +1.40m ⚠️ |
| dist/clipboard.976debcc.js | 177.00b | +1.00b ⚠️ | 45.38s | +5.44s ⚠️ |
| dist/browser.976debcc.js | 177.00b | +1.00b ⚠️ | 1.42m | +44.98s ⚠️ |
| dist/editorView.976debcc.js | 177.00b | -1.00b 🚀 | 44.75s | -16.51s 🚀 |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | -5.00b 🚀 | 47.01s | -1.26m 🚀 |
| dist/16.976debcc.js | 174.00b | -5.00b 🚀 | 45.38s | +3.94s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 2.33mb | +6.00b ⚠️ | 394.00ms | +14.00ms |
| dist/Toolbar.5c7a6fbb.js | 105.62kb | +0.00b | 137.00ms | +7.00ms ⚠️ |
| dist/workerHasher.d7e20c98.js | 11.56kb | +0.00b | 140.00ms | +7.00ms |
| dist/index.css | 3.38kb | +0.00b | 114.00ms | +9.00ms ⚠️ |
| dist/workerHasher.854262ce.js | 1.48kb | +0.00b | 140.00ms | +7.00ms |
| dist/link.6fa27bb2.js | 1.20kb | +0.00b | 149.00ms | +8.00ms ⚠️ |
| dist/emoji.c4bc5355.js | 1.12kb | +0.00b | 155.00ms | +11.00ms ⚠️ |
| dist/simpleHasher.884d4eb6.js | 446.00b | +0.00b | 140.00ms | +7.00ms |
| dist/esm.976debcc.js | 176.00b | +2.00b ⚠️ | 166.00ms | +6.00ms |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | +2.00b ⚠️ | 165.00ms | +7.00ms |
| dist/dropzone.976debcc.js | 176.00b | +2.00b ⚠️ | 161.00ms | +7.00ms |
| dist/clipboard.976debcc.js | 176.00b | +2.00b ⚠️ | 160.00ms | +6.00ms |
| dist/browser.976debcc.js | 176.00b | +2.00b ⚠️ | 160.00ms | +7.00ms |
| dist/16.976debcc.js | 176.00b | +2.00b ⚠️ | 151.00ms | +6.00ms |
| dist/editorView.976debcc.js | 176.00b | +2.00b ⚠️ | 130.00ms | +7.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/254772734773428737)**

File: packages/core/core/src/AssetGraph.js [hunk 1]
```diff
@@ -36,7 +37,7 @@ type SerializedAssetGraph = {|
   hash: ?string,
 |};
 
-const nodeFromDep = (dep: Dependency): DependencyNode => ({
+export const nodeFromDep = (dep: Dependency): DependencyNode => ({
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: nit: can this be a function declaration?

File: packages/core/core/src/AssetGraph.js [hunk 2]
```diff
@@ -209,17 +211,53 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
...
+      this.resolveAsset(assetNode, dependentAssets);
+    }
+  }
+
+  resolveAsset(assetNode: AssetNode, dependentAssets: Array<Asset>) {
+    let depNodes = [];
+    let depNodesWithAssets = [];
+    for (let dep of assetNode.value.dependencies.values()) {
+      let depNode = nodeFromDep(dep);
+      depNodes.push(this.nodes.get(depNode.id) || depNode);
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: nit: `??`

File: packages/core/core/src/AssetGraph.js [hunk 3]
```diff
@@ -209,17 +211,53 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
       return;
     }
 
-    let assetNodes = assets.map(asset => nodeFromAsset(asset));
-    this.replaceNodesConnectedTo(assetGroupNode, assetNodes);
+    let dependentAssetKeys = [];
+    let assetTuples: Array<[AssetNode, Array<Asset>, boolean]> = [];
+    for (let asset of assets) {
+      let isDirect = dependentAssetKeys.includes(asset.uniqueKey)
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: nit: `let isDirect = !dependentAssetKeys.includes(asset.uniqueKey);`

File: packages/core/core/src/AssetGraph.js [hunk 4]
```diff
@@ -209,17 +211,53 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
...
 
-    for (let assetNode of assetNodes) {
-      let depNodes = [];
-      invariant(assetNode.type === 'asset');
-      for (let dep of assetNode.value.dependencies.values()) {
-        let depNode = nodeFromDep(dep);
-        depNodes.push(this.nodes.get(depNode.id) || depNode);
+    this.replaceNodesConnectedTo(
+      assetGroupNode,
+      assetTuples.filter(a => a[2]).map(a => a[0]),
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: could these tuples be objects with named properties rather than relying on position?

**Pull request merged**